### PR TITLE
feat: low level read API

### PR DESCRIPTION
**Pull request by julien-f**

### CODE REVIEW

File: lib/api/open.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+var request = require('../tools/smb2-forge').request;
+
+module.exports = function open(path, cb) {
```
Comment (marsaud) [inline lib/api/open.js:3]: openFile ?...

Comment (julien-f) [reply lib/api/open.js:3]: This method is the match of `fs.open` (like `read`) but I couldn't use the name `close` due to it being taken by the method closing the connection.

I'm not against changing the name but then we might wonder about `read` as well…

File: lib/api/read.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    offset += n;
+    position = position.add(n);
+
+    readChunk();
+  }
+  function readChunk() {
+    if (bytesRead >= length || position.ge(fileSize)) {
+      cb(null, buffer, bytesRead);
+    }
+    request(
```
Comment (marsaud) [inline lib/api/read.js]: Do you really intend to call read one last time after the previous test called the final cb ?

Comment (julien-f) [reply lib/api/read.js]: Nope, that's a mistake, thank you.

File: lib/structures/constants.js [hunk 3]
```diff
@@ -9,4 +9,7 @@ module.exports = {
   FILE_OPEN_IF: 0x00000003,
   FILE_OVERWRITE: 0x00000004,
   FILE_OVERWRITE_IF: 0x00000005,
+
+  // Where does it come from?!
```
Comment (marsaud) [inline lib/structures/constants.js:13]: This value was provided as is when I discovered the package, I did not question it.

Comment (julien-f) [reply lib/structures/constants.js:13]: Yes, that's what I thought, but I think it makes sense to keep the comment as long as it's unclear where this magic number comes from, don't you agree?

File: lib/smb2.js [hunk 4]
```diff
@@ -100,3 +100,7 @@ proto.mkdir = SMB2Connection.requireConnect(require('./api/mkdir'));
 
 proto.ensureDir = SMB2Connection.requireConnect(require('./api/ensureDir'));
 proto.getSize = SMB2Connection.requireConnect(require('./api/getSize'));
+
+proto.open = SMB2Connection.requireConnect(require('./api/open'));
```
Comment (marsaud) [inline lib/smb2.js:104]: The requireConnect may schedule an auto-close so subsequent reads and/or closeFile may crash, unless connection.autoCloseTimeout is 0.

Comment (julien-f) [reply lib/smb2.js:104]: Yes, I should put a note in the README.

File: lib/api/open.js (hunk 1 above)
Comment (marsaud) [reply lib/api/open.js:3]: And there is a readFile somewhere already, if I remember well, hope this won't collide on the same proto...

File: lib/structures/constants.js (hunk 3 above)
Comment (marsaud) [reply lib/structures/constants.js:13]: I agree.

File: lib/smb2.js (hunk 4 above)
Comment (marsaud) [reply lib/smb2.js:104]: What do you think of switching to 0 as the default value for a close future release ?

File: lib/api/open.js (hunk 1 above)
Comment (julien-f) [reply lib/api/open.js:3]: `readfile` is already use to read the whole file (and that's also the case in `fs.readFile`).

**Review by marsaud — APPROVED** [review]
Looks like a common agreement.

File: lib/structures/constants.js (hunk 3 above)
Comment (tracker1) [reply lib/structures/constants.js:13]: FYI: https://msdn.microsoft.com/en-us/library/cc246502.aspx

CreateDisposition option...

Comment (julien-f) [reply lib/structures/constants.js:13]: Can you explain a bit more?

I guess it's the max size of a request packet minus the other fields, but i can't find it in the page :confused:

Comment (tracker1) [reply lib/structures/constants.js:13]: Sorry... that's something else, I didn't see the line below the comment... lol... the SMB max read size seems to be 65536 from a google search[1]

My bad on the reference... lol

[1] https://social.technet.microsoft.com/Forums/windowsserver/en-US/83ab5d24-4815-4161-8baa-5cd9c6e98506/smb2-read-size-limited-to-65536-bytes?forum=winserverfiles

**Pull request merged**